google-cloud-spanner/src/main/java/com/google/cloud/spanner/ForwardingAsyncResultSet.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner;
+
+import com.google.api.core.ApiFuture;
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import java.util.concurrent.Executor;
+
+/** Forwarding implementation of {@link AsyncResultSet} that forwards all calls to a delegate. */
+public class ForwardingAsyncResultSet extends ForwardingResultSet implements AsyncResultSet {
```

Comment:
It's used by [read/write transactions](https://github.com/googleapis/java-spanner/blob/a7dd1dd3376784cf24e8aeb64a7155e26b9192d5/google-cloud-spanner/src/main/java/com/google/cloud/spanner/TransactionRunnerImpl.java#L95) to prevent the `TransactionRunner` from committing the transaction before all `AsyncResultSet`s that have been created by the transaction have returned `DONE` from the `cursorReady` method.

Consider [this](https://github.com/googleapis/java-spanner/blob/a7dd1dd3376784cf24e8aeb64a7155e26b9192d5/google-cloud-spanner/src/test/java/com/google/cloud/spanner/AsyncRunnerTest.java#L599) test case. It roughly executes the following flow:
1. Start an asynchronous read/write transaction.
1. Execute an asynchronous query in the read/write transaction and consume the result set using a background executor.
1. The `doWorkAsync` method finishes directly after starting the async query. The `AsyncResultSet` will at that moment probably still be processed by the background executor.
1. The underlying `TransactionContext` keeps track of the `AsyncResultSet`s that have been opened, and will not commit the transaction until all `AsyncResultSet`s and other async statements have finished.

Reply:
I see, thanks for the thorough response.